src/Libraries/DSCPython/DCPythonCodeCompletionProviderCore.cs

```diff
@@ -0,0 +1,806 @@
+﻿using Autodesk.DesignScript.Interfaces;
+using Dynamo.Logging;
+using Dynamo.PythonServices;
+using Python.Runtime;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+
+namespace DSCPython
+{
+    internal class DSCPythonCodeCompletionProviderCore : IExternalCodeCompletionProviderCore, ILogSource, IDisposable
+    {
+        #region Private members
+        /// <summary>
+        /// Maps a basic variable regex to a basic python type.
+        /// </summary>
+        private List<Tuple<Regex, Type>> BasicVariableTypes;
+
+        /// <summary>
+        /// Tracks already referenced CLR modules
+        /// </summary>
+        private HashSet<string> ClrModules { get; set; }
+
+        /// <summary>
+        /// Keeps track of failed statements to avoid poluting the log
+        /// </summary>
+        private Dictionary<string, int> BadStatements { get; set; }
+
+        private PythonEngine engine;
+        /// <summary>
+        /// The engine used for autocompletion.  This essentially keeps
+        /// track of the state of the editor, allowing access to variable types and
+        /// imported symbols.
+        /// </summary>
+        private object Engine
+        {
+            get { return engine; }
+            set { engine = (PythonEngine)value; }
+        }
+        private PyScope scope;
+
+        private string uniquePythonScopeName = "uniqueScope";
+
+        /// <summary>
+        /// The scope used by the engine.  This is where all the loaded symbols
+        /// are stored.  It's essentially an environment dictionary.
+        /// </summary>
+        private PyScope Scope
+        {
+            get { return scope; }
+            set { scope = (PyScope)value; }
+        }
+        #endregion
+
+        #region Private Methods
+        private PyScope CreateUniquePythonScope()
+        {
+            PyScopeManager.Global.TryGet(uniquePythonScopeName, out PyScope Scope);
+
+            if (Scope == null)
+            {
+                Scope = Py.CreateScope(uniquePythonScopeName);
+                Scope.Exec(@"
+import clr
+import sys
+clr.setPreload(True)
+");
+            }
+
+            return Scope;
+        }
+
+        private object ExecutePythonScriptCode(string code)
+        {
+            Python.Included.Installer.SetupPython().Wait();
+
+            if (!PythonEngine.IsInitialized)
+            {
+                PythonEngine.Initialize();
+                PythonEngine.BeginAllowThreads();
+            }
+
+            IntPtr gs = PythonEngine.AcquireLock();
+            try
+            {
+                using (Py.GIL())
+                {
+                    var result = Scope.Eval(code);
+                    return CPythonEvaluator.OutputMarshaler.Marshal(result);
```

Comment:
Using the Evaluator OutputMarshaler might not be the best thing.
I am thinking of either adding a dedicated Marshaler or trying to re use the CPythonEvaluator.Evaluate method directly (probably too much refactor)